### ISSUE TITLE
Add Hive EV Charger

### ISSUE THREAD

**Issue by drt102** [Pyhass/Pyhiveapi#43]
**Is your feature request related to a problem? Please describe.**
Is it possible to add in the Hive EV Charger

**Describe the solution you'd like**
I use Home Assistant and I'd like to be able to monitor and control my Hive EV Charger.

**Comment by KJonline** [comment]
@drt102 yes it would be possible is my presumption. Unfortunately even though I have an electric car I live in a terraced house so I can’t get the ev charger yet. If you would be happy to share some of your hive api data with me we can have a look at how to develop this?

**Comment by KJonline** [comment]
@drt102 the easiest way is to login to the hive website with a browser and open up the debug console which I think is right click and inspect on chrome. Are you familiar with the tools?


From their if you go to the network tab and filter by XHR requests and reload the hive dashboard page you should see some requests called products and devices they are 2 separate requests.

If you could please extract the responses and share them we can have a look at the code. They may be quite large depending on how many hive products you have.

Please make sure you review the responses yourself before sharing and remove any identifiable information to you personally.

**Comment by drt102** [comment]
I've extracted just the bits for the EV charger and masked out the individual IDs

**Device**
[device_masked.txt](https://github.com/Pyhass/Pyhiveapi/files/8912993/device_masked.txt)

**Product**
[product._masked.txt](https://github.com/Pyhass/Pyhiveapi/files/8912997/product._masked.txt)

**Comment by KJonline** [comment]
Thanks for that I can work with it. What actions can you take when you click on the ev charger in the app? I’m guessing turn it on and turn it off are there any other actions?

also I can some additional data which could add some sensors e.g plugged and miles added.

Next step now would be to do the same extract when the vehicle is charging and if there is and on and off action we need to capture the api call being made when that is called. 

All this can be achieved using the same mechanism to get the data as you already have done

**Comment by drt102** [comment]
My EV is not quite here yet (hopefully in next 2 weeks) so I haven't actually had the charger working yet but the config options are as follows:
1. Charger - Eco-charging option to schedule charging to take place between 12am and 8am.
2. Tariff - Chose your electricity supplier from a list and pick a single or multi-rate tariff and set the unit price
3. Vehicle - Enter vehicle efficiency in miles per kWh
4. Notifications - Charging started, charging completed, charging reminders
On the main screen is what looks like a button to override any schedules and force charging to start
The charger is therefore not very smart but I was hoping that I would be able to trigger the charging to start through HA.

**Comment by KJonline** [comment]
Yep so when viewing the item on the hive website can you do the same as before with the debug tools in the browser out this time click the button to start charging now. This should show an api call to change the state we will need an example of that

**Comment by drt102** [comment]
The Hive website doesn't show any of this for the EV charger - I just get a blank screen when I click on the device from the dashboard. I can only see all of the controls and options from the app.

**Comment by KJonline** [comment]
Ah ok, it can still be done. Are you familiar with http proxy’s and proxying your mobile device traffic through one?

**Comment by drt102** [comment]
I've now get Charles Proxy set up and have tested recording traffic from my mobile. What do you need me to do to progress work on getting the Hive EV Charger set up?

**Comment by KJonline** [comment]
@drt102 have you gone through the process of install the certificate on your device so you can read ssl encrypted traffic. 

Once you have done that and you are happy you can read and see the hive api calls.

first thing to start of with is record some api calls being made. So set recording going and then use the EV option in the hive app try and cover all functionality in it so changing mode if there is one turning it on and off etc 

once you have done that as a start just copy the urls for hive endpoints into this chat.

there shouldn’t be but please double check there is no personal identifiable information. If there is just put a placeholder in the data e.g <name>

**Comment by drt102** [comment]
Hmm, I think there is an issue with Android App trying to use Charles Proxy:

_Android
As of Android N, you need to add configuration to your app in order to have it trust the SSL certificates generated by Charles SSL Proxying. This means that you can only use SSL Proxying with apps that you control._

Does this mean that I can't use this with the Hive Android App?

**Comment by KJonline** [comment]
@drt102  yes reading that it seems it won’t work unless you are developing the app

**Comment by jameslegg** [comment]
I had a crack at getting this working with my Hive charger using some captures from ProxyMan on iOS + Claude.

I'm using a Hive branded Alfen Charger, which I don't _think_ Britsh Gas fit anymore, though it's only a year or so old. It's possible all the current chargers use the same endpoint (which appears to be different from the heating) but I can't test.

I've got a branch over here https://github.com/jameslegg/Pyhiveapi/tree/feature/evse-support 

It's 100% AI coded after handing it the captures, though I did QA it manually to ensure it works as expected and had it fix up a few things that needed it. I've been using it for my use case of comparing the car's charges status against the charger and alerting me when they disagree for a week or so now. 

I'd be happy to submit a PR, but I don't want to inflict the burden of reviewing the AI output on somebody else - I think it's sane, but I'm absolutely not a python professional.
